File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"errors"
	"net"
)

const MAX_STR_SIZE = 255
const DOCUMENT_SIZE = 4
const NUMBER_SIZE = 2
const ANSWER_SIZE = 1
const BATCH_SIZE = 2
const MAX_BATCH_BYTES = 8000 // 8KB

const SUCESS = 0
const FAIL = 1
/// Read all the bytes from the connection to avoid partial reads
func RecvAll(conn net.Conn, size int) ([]byte, error) {
    buf := make([]byte, size)
    total := 0

    for total < size {
        n, err := conn.Read(buf[total:])
        if err != nil || n == 0 {
            return nil, err
        }
        total += n
    }
    return buf, nil
}

// Send all the bytes from the message to avoid partial writes
func send_all(conn net.Conn, message []byte) error{
	written := 0 
	for written < len(message) {
		n, err := conn.Write(message)
				if err != nil || n == 0 {
			log.Criticalf(
				"action: send_all | result: fail| total read: %v |  error: %v",
				err,
			)
			return err
		}
		written += n
	}
	return nil
}
// Serialize a string with unknown size, the first byte is the size of the string
func serializeUnknownString(message string, buf []byte) []byte{
	if len(message) > MAX_STR_SIZE {
		log.Criticalf( 
			"action: serialize_unknown_string | result: fail | error: string too long",
		)
		return nil
	}
	buf = append(buf, byte(len(message)))
	buf = append(buf, []byte(message)...)
	return buf
}


func SendBet(conn net.Conn, bet *Bet) error {
	msg, err := encodeBet(bet)
	if err != nil {return err}
	err = send_all(conn, msg)
	if err != nil {return err} else {return nil}
}

func encodeBet (bet *Bet) ([]byte, error) {
	// agency
	msg := make([]byte, 0)
	msg = append(msg, bet.agency)

	// firstName
	msg = serializeUnknownString(bet.firstName, msg)
	if msg == nil {return nil, errors.New("error serializing firstName")}

	// lastName
	msg = serializeUnknownString(bet.lastName, msg)
	if msg == nil {return nil, errors.New("error serializing lastName")}

	// document
	docBytes := make([]byte, DOCUMENT_SIZE)
	binary.BigEndian.PutUint32(docBytes, bet.document)
	msg = append(msg, docBytes...)

	// birthDate
	msg = append(msg, []byte(bet.birthDate)...) // SIZE 10

	// number
	numBytes := make([]byte, NUMBER_SIZE)
	binary.BigEndian.PutUint16(numBytes, bet.number)
	msg = append(msg, numBytes...)

	return msg, nil 
}
// Send a batch of bets to the server
// The first two bytes are the number of bets
// Then it sends the bets
func SendBatch(conn net.Conn, bets []*Bet) error {
	var batches [][]byte
	var currentBatch []byte

	// Add total number of bets at the beginning of the first batch
	currentBatch = make([]byte, BATCH_SIZE) // Initialize with 2 bytes
	binary.BigEndian.PutUint16(currentBatch, uint16(len(bets)))

	for _, bet := range bets {
		betMsg, err := encodeBet(bet)
		if err != nil {
			return err
		}

		// If adding betMsg exceeds the max batch size, flush the current batch
		if len(currentBatch)+len(betMsg) > MAX_BATCH_BYTES {
			batches = append(batches, currentBatch) // Save current batch
			currentBatch = make([]byte, 0)          // Start a new empty batch (no bet count in subsequent batches)
		}

		// Append the bet message to the current batch
		currentBatch = append(currentBatch, betMsg...)
	}

	// Add the last batch if there's any remaining data
	if len(currentBatch) > 0 {
		batches = append(batches, currentBatch)
	}

	// Send all batches
	for _, batch := range batches {
		if err := send_all(conn, batch); err != nil {
			return err
		}
	}

	return nil
}

func RecvAnswer(conn net.Conn) (int,error) {
	answer, err := RecvAll(conn, ANSWER_SIZE)
	if err != nil {return -1, err}
	answer_v := int(answer[0])			
	return answer_v, nil 
}

File: server/common/protocol.py
import socket
from common.utils import Bet

AGENCY_SIZE = 1 
STR_SIZE = 1 
DOCUMENT_SIZE = 4 
BIRTHDATE_SIZE = 10 
NUMBER_SIZE = 2 
MAX_STR_SIZE = 255
ANSWER_SIZE = 1

## Server answers
SUCESS = 0
ERROR_GENERIC = 1

"""
Reads a bet from the socket and returns it.
The bet is expected to be in the following format:
    - 1 byte for the agency
    - 1 byte for the first name size
    - (0-255) bytes for the first name
    - 1 byte for the last name size
    - (0-255) bytes for the last name
    - 4 bytes for the document
    - 10 bytes for the birthdate
    - 2 bytes for the number

"""

def recv_bet(skt: socket.socket) -> Bet:
    data = recv_all(skt, AGENCY_SIZE)
    agency = int.from_bytes(data, byteorder='big')
    data = recv_str_uknown_size(skt)
    first_name = data
    data = recv_str_uknown_size(skt)
    last_name = data
    data = recv_all(skt, DOCUMENT_SIZE)
    document = int.from_bytes(data, byteorder='big')
    data = recv_all(skt, BIRTHDATE_SIZE) 
    birthdate = data.decode('utf-8')

    data = recv_all(skt, NUMBER_SIZE)
    number = int.from_bytes(data, byteorder='big')

    return Bet(agency, first_name, last_name, document, birthdate, number)

"""
reads a string from the socket with an unknown size, the size is read first
"""
def recv_str_uknown_size(skt: socket.socket) -> str:
    data = recv_all(skt, STR_SIZE)
    size = int.from_bytes(data, byteorder='big')
    if size > MAX_STR_SIZE:
        raise ValueError("String size exceeded") 
    data = recv_all(skt, size)
    return data.decode('utf-8')


"""
reads the entire message from the socket to avoid short-reads
"""
def recv_all(skt: socket.socket, size: int) -> bytes:
    data = b'' 
    while len(data) < size:
        pkt = skt.recv(size - len(data))
        if len(pkt) == 0:
            raise OSError("Connection closed")
        data += pkt

    return data

"""
sends the entire message to the socket to avoid short-writes
"""
def send_all(skt: socket.socket, data: bytes) -> None: 
    while len(data) > 0:
        sent = skt.send(data)
        if sent == 0:
            raise OSError("Connection closed")
        data = data[sent:]


def send_answer(skt: socket.socket, answer: int) -> None:
    if answer not in [SUCESS, ERROR_GENERIC]:
        raise ValueError("Invalid answer")
    data = answer.to_bytes(ANSWER_SIZE, byteorder='big')
    send_all(skt, data)